### ISSUE TITLE
Unable to discover bacnet devices with bacnet-client application running inside docker container

### ISSUE THREAD

**Issue by saransh14** [Code-House/bacnet4j-wrapper#22]
I'm trying to discover the BACnet devices [simulated devices, using YABE for same] via the java application internally using the bacnet4j-wrapper.
<img width="982" alt="successful-logs" src="https://github.com/Code-House/bacnet4j-wrapper/assets/21282950/1f3bebc8-02ca-4806-85dc-4b48ff242bba">

Devices are getting discovered when running the application on local system, but when we're trying to run the code inside the docker container, the client application is unable to discover the devices.
![docker-container-logs](https://github.com/Code-House/bacnet4j-wrapper/assets/21282950/2c3f6908-6305-4983-91ad-d5c5bbc680a8)

I've even tried using the --network="host" but no luck.

**Comment by splatch** [comment]
Hello, for docker issue comes from visibility of networks, try with `--network host`.

**Comment by saransh14** [comment]
Hi, I created a java application using bacnet4j-wrapper, which discovers the bacnet devices and fetches all the objects which  are part of the devices and eventually fetches the present-value using the BacNetIpClient.

NOTE: I'm using YABE simulator [Bacnet.Room.Simulator.exe] for simulating a BACnet device

I'm able to discover the devices when running the java application on my local system. 

Following are the code snippets:
<img width="827" alt="code-snip-1" src="https://github.com/Code-House/bacnet4j-wrapper/assets/21282950/3ce52ea6-2b45-433c-af4c-6192787c0319">
<img width="923" alt="code-snip-2" src="https://github.com/Code-House/bacnet4j-wrapper/assets/21282950/4fe5f16b-7a76-41a0-97a4-74b7870d645a">


Following are the logs attached [when running the java application locallly]:
<img width="982" alt="successful-logs" src="https://github.com/Code-House/bacnet4j-wrapper/assets/21282950/ae1d170b-092f-495c-a461-2f27a740ac80">


Wireshark analysis also if perfect in this case:
<img width="1508" alt="wireshark-1" src="https://github.com/Code-House/bacnet4j-wrapper/assets/21282950/171ff171-e722-4b17-ab7a-d2ea40e6d860">


Now I tried running the above java application inside a docker container with --network="host", I'm able to see the who-is request in wireshark, and infact the i-am been responded by the simulated device too, but the application is not able to listen to the i-am response and thus not discovering it.

Wireshark analysis [when running bacnet-client inside docker container with --network="host"]
![wireshark-2](https://github.com/Code-House/bacnet4j-wrapper/assets/21282950/598f3618-cce3-49ec-9ba7-4cd3a2739bc3)

**Comment by splatch** [comment]
Network trace from last screenshot indicate that there are two IP addresses used, 127.0.0.1 and 192.168.29.30. Your machine IP is probably 192.168.29.96. Can you check with other variant of constructor - `BacNetIpClient(String ip, String broadcast, int deviceId)` and see if it improves situation? The first ip should be your docker host which is visible to bacnet devices. Also are you running host using windows or linux?

**Comment by saransh14** [comment]
YABE Device simulator is running on 192.168.29.30. 
I was initially running the BACnet-client Java application on my local setup [macOS, IP: 192.168.29.96]. This belongs to the success scenario.


Now I'm trying to run the BACnet-client Java application as docker container on my linux machine [IP: 192.168.29.30]
Device simulator is also running on same linux machine [IP: 192.168.29.30]

NOTE: Docker is installed on linux machine [IP: 192.168.29.30], and YABE simulator is also running on same linux machine [IP: 192.168.29.30]

IP address 127.0.0.1 is the one sending the who-is request [BACnet-client docker container, IP is 127.0.0.1 because I'm using --network="host"]



When using other variant of constructor - BacNetIpClient(String ip, String broadcast, int deviceId) and defining values as BacNetIpClient("192.168.29.30", "255.255.255.255", "2002"), I'm get BindException (attaching the log trace).
![bind-exception](https://github.com/Code-House/bacnet4j-wrapper/assets/21282950/a89b3f15-fafa-4086-85d6-0d047f862b8f)

**Comment by splatch** [comment]
The `--network=host` indicates no network isolation, so your container can access host network interfaces. Namely it removes constraints and extra hop in routing the traffic. This means that you could actually bind any of host IP. When you run two processes which are trying to listen/broadcast bacnet packets they will end up fighting same network port. I usually ran simulator on another box to avoid conflicts.
Anyhow - try broadcast address of `192.168.29.255`. Making `255.255.255.255` will likely require root permissions.

**Comment by saransh14** [comment]
When using other variant of constructor - BacNetIpClient(String ip, String broadcast, int deviceId) and defining values as BacNetIpClient("192.168.29.30", "192.168.29.255", "2002"), I'm getting the same BindException.

And when trying BacNetIpClient(String broadcast, int deviceId) and defining values as BacNetIpClient("192.168.29.255", "2002"), I'm unable to discover devices inside docker container and who-is requests is also not captured under wireshark
<img width="798" alt="image" src="https://github.com/Code-House/bacnet4j-wrapper/assets/21282950/8c7c718e-0eb3-4d93-83de-15e829caf068">

**Comment by splatch** [comment]
I've created a basic docker image, see #23.

Running it I can see devices in my host network: `docker run --rm --net=host code-house/bacnet4j-wrapper:1.3.0-SNAPSHOT 192.168.2.255`. Image uses `DiscoveryMain` class which you can find in project sources.

**Comment by saransh14** [comment]
I've pushed my bacnet-client code onto github along with the Dockerfile that I've created.
Link: https://github.com/saransh14/BACnet-Exploration

Can you please give it a try on your setup by dockerising and running the dockerimage as `docker run --net host <image-name>`

**Comment by splatch** [comment]
For me `docker run --rm --net=host -e BROADCAST_ADDRESS=192.168.2.255 saransh14/bacnet-exploration` did the trick. I am running linux host.

**Comment by saransh14** [comment]
I'm still unable to discover the devices inside docker container.
Now even I've ran the YABE simulator on another windows machine under same wifi-network, and bacnet-client as docker container on linux host machine

Are you using the bindIP same as "0.0.0.0", as specified in code base, while creating object for `BacNetIpClient`?

**Comment by splatch** [comment]
Yes, the above docker execution from your repo used your code. Only one parameter I tweaked was broadcast address. Anyhow, if you enumerate network interfaces using standard Java APIs you will get from it also a broadcast address which you can use to setup transport.

**Comment by saransh14** [comment]
Thanks @splatch for your support, there was some network interface issue.

I'm now able to discover the BACnet devices under the docker environment using `--network=host`.
I ran the YABE simulator on one linux machine and docker setup for bacnet-client application on another linux host machine so that both don't try to listen/broadcast bacnet packets and end up fighting same network port.
Post then I was able to discover the BACnet device which were present under the same subnet.

Thanks for your support, closing this issue.
